File: src/main/java/hu/frontrider/arcana/creatureenchant/effect/EffectEnchantBase.java
package hu.frontrider.arcana.creatureenchant.effect;

import hu.frontrider.arcana.creatureenchant.backend.CreatureEnchant;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.event.entity.living.LivingEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

public abstract class EffectEnchantBase extends CreatureEnchant {

    private final Potion positive;
    private final Potion negative;
    private final double negativeMultiplier;

    public EffectEnchantBase(Potion positive, Potion negative, double negativeMultiplier, ResourceLocation resourceLocation, String unlocalisedName) {
        super(resourceLocation,unlocalisedName);
        this.positive = positive;
        this.negative = negative;
        this.negativeMultiplier = negativeMultiplier;
    }

    @SubscribeEvent
    public void tick(LivingEvent.LivingUpdateEvent event) {

        EntityLivingBase entity = event.getEntityLiving();
        int enchantLevel = getEnchantLevel(entity, this);

        if (enchantLevel > 0) {
            if(entity.getActivePotionEffect(positive) == null || entity.getActivePotionEffect(positive).getDuration() <30)
            entity.addPotionEffect(new PotionEffect(positive,240,enchantLevel-1,true,false));
        }

        if (enchantLevel < 0) {
            if(entity.getActivePotionEffect(negative) == null || entity.getActivePotionEffect(negative).getDuration() <30)
            entity.addPotionEffect(new PotionEffect(negative,240, (int) ((enchantLevel+1)*negativeMultiplier),true,false));
        }
    }

}

File: src/main/java/hu/frontrider/arcana/core/creatureenchant/EffectEnchantBase.kt
package hu.frontrider.arcana.core.creatureenchant

import net.minecraft.potion.Potion
import net.minecraft.potion.PotionEffect
import net.minecraft.util.ResourceLocation
import net.minecraftforge.event.entity.living.LivingEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

abstract class EffectEnchantBase(private val positive: Potion, private val negative: Potion, private val negativeMultiplier: Double,resourceLocation: ResourceLocation, unlocalisedName: String) : CreatureEnchant(resourceLocation, unlocalisedName) {

    constructor(positive: Potion,negative: Potion,negativeMultiplier: Double, resourceLocation: ResourceLocation):this(positive,negative,negativeMultiplier,resourceLocation,resourceLocation.resourcePath)

    @SubscribeEvent
    open fun tick(event: LivingEvent.LivingUpdateEvent) {

        val entity = event.entityLiving
        val enchantLevel = getEnchantLevel(entity, this)

        if (enchantLevel > 0) {
            if (entity.getActivePotionEffect(positive) == null || entity.getActivePotionEffect(positive)!!.duration < 30)
                entity.addPotionEffect(PotionEffect(positive, 240, enchantLevel - 1, true, false))
        }

        if (enchantLevel < 0) {
            if (entity.getActivePotionEffect(negative) == null || entity.getActivePotionEffect(negative)!!.duration < 30)
                entity.addPotionEffect(PotionEffect(negative, 240, ((enchantLevel + 1) * negativeMultiplier).toInt(), true, false))
        }
    }

}
